ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,284 @@
+    // add the operation representation to the list of traces
+    op.setVmOperationExecutionReport(report);
+    currentTrace.add(op);
+    currentIndex++;
+    lastCapturedStorage = currentTraceFrame.getStorage();
+  }
+
+  private void handleDepthIncreased(final VmOperation op, final VmOperationExecutionReport report) {
+    // check if next frame depth has increased i.e the current operation is a call
+    if (maybeNextFrame.map(next -> next.isDeeperThan(currentTraceFrame)).orElse(false)) {
+      maybeNextFrame.ifPresent(
+          nextFrame -> op.setCost(nextFrame.getGasRemaining().toLong() + op.getCost()));
+      final VmTrace newSubTrace = new VmTrace();
+      parentTraces.addLast(newSubTrace);
+      findLastFrameInCall(currentTraceFrame, currentIndex)
+          .ifPresent(
+              lastFrameInCall -> {
+                report.setUsed(lastFrameInCall.getGasRemaining().toLong());
+                lastFrameInCall
+                    .getStack()
+                    .filter(stack -> stack.length > 0)
+                    .map(stack -> stack[stack.length - 1])
+                    .map(last -> Quantity.create(UInt256.fromHexString(last.getHexString())))
+                    .ifPresent(report::addPush);
+              });
+      op.setSub(newSubTrace);
+    }
+  }
+
+  private void handleDepthDecreased() {
+    // check if next frame depth has decreased i.e the current operation closes the parent trace
+    if (maybeNextFrame.map(next -> next.isLessDeepThan(currentTraceFrame)).orElse(false)) {
+      currentTrace = parentTraces.removeLast();
+    }
+  }
+
+  private VmOperation buildVmOperation() {
+    final VmOperation op = new VmOperation();
+    // set gas cost and program counter
+    op.setCost(currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    op.setPc(currentTraceFrame.getPc());
+    return op;
+  }
+
+  private VmOperationExecutionReport generateExecutionReport() {
+    final VmOperationExecutionReport report = new VmOperationExecutionReport();
+    // set gas remaining
+    report.setUsed(
+        currentTraceFrame.getGasRemaining().toLong()
+            - currentTraceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    return report;
+  }
+
+  private void generateTracingMemory(final VmOperationExecutionReport report) {
+    if (currentTraceFrame.getDepth() == maybeNextFrame.map(TraceFrame::getDepth).orElse(0)) {
+      updatedMemory(currentTraceFrame.getMemory(), maybeNextFrame.flatMap(TraceFrame::getMemory))
+          .filter(memory -> memory.length > 0)
+          .map(TracingUtils::dumpMemoryAndTrimTrailingZeros)
+          .map(Mem::new)
+          .ifPresent(report::setMem);
+    }
+  }
+
+  private void generateTracingPush(final VmOperationExecutionReport report) {
+    // set push from stack elements if some elements have been produced
+    if (currentTraceFrame.getStackItemsProduced() > 0 && maybeNextFrame.isPresent()) {
```

Comment:
What happens if the last operation pushes items to the stack?  It looks like we'll skip reporting the stack changes in that case.  Can we add a test for this?

Reply:
I have updated a test to cover this case and it works.
the bycode is `0x6020356000355560603560403555600000`
Decoded Bytecode:
[1] PUSH1 0x20
[2] CALLDATALOAD
[4] PUSH1 0x00
[5] CALLDATALOAD
[6] SSTORE
[8] PUSH1 0x60
[9] CALLDATALOAD
[11] PUSH1 0x40
[12] CALLDATALOAD
[13] SSTORE
[15] PUSH1 0x00
[16] STOP

The last operation pushes item to the stack.

Reply:
Oh there is a `STOP`, you probably mean without a `STOP`

Reply:
Yeah - does it work without the explicit STOP?

Reply:
It should. I have removed the `maybeNextFrame.isPresent` condition. More generally i don't use `maybeNextFrame` anymore.

Reply:
I will test it.